src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractMergeIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,91 @@
+
+        protected abstract Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetChangedRoot(
+            SyntaxNode root, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, SyntaxGenerator generator);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            if (IsApplicableSpan(node, context.Span, out var ifStatement))
+            {
+                var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+
+                if (await CanBeMergedAsync(context.Document, ifStatement, syntaxFacts, context.CancellationToken).ConfigureAwait(false))
+                {
+                    context.RegisterRefactoring(
+                        CreateCodeAction(
+                            c => RefactorAsync(context.Document, context.Span, syntaxFacts, c)));
+                }
+            }
+        }
+
+        private async Task<Document> RefactorAsync(Document document, TextSpan span, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
+
+            Contract.ThrowIfFalse(IsApplicableSpan(node, span, out var ifStatement));
+
+            var newRoot = GetChangedRoot(root, ifStatement, syntaxFacts, document.GetLanguageService<SyntaxGenerator>());
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected static IReadOnlyList<SyntaxNode> WalkDownBlocks(ISyntaxFactsService syntaxFacts, IReadOnlyList<SyntaxNode> statements)
+        {
+            // If our statements only contain a single block, walk down the block and any subsequent nested blocks
+            // to get the real statements inside.
+
+            while (statements.Count == 1 && syntaxFacts.IsPureBlock(statements[0]))
+            {
+                statements = syntaxFacts.GetExecutableBlockStatements(statements[0]);
+            }
+
+            return statements;
+        }
+
+        protected static IReadOnlyList<SyntaxNode> WalkUpBlocks(ISyntaxFactsService syntaxFacts, IReadOnlyList<SyntaxNode> statements)
+        {
+            // If our statements are inside a block, walk up the block and any subsequent nested blocks that contain
+            // no other statements to get the topmost block.
+
+            while (statements.Count > 0 && syntaxFacts.IsPureBlock(statements[0].Parent) &&
+                   syntaxFacts.GetExecutableBlockStatements(statements[0].Parent).Count == statements.Count)
```

Comment:
when woudl the third condition not be true?

Reply:
If there are statements below the block.

Reply:
For example:
```c#
if (condition)
{
    {
        OurStatement();
    }
    AnotherStatement();
}
```